To allow institution-specific setup

biocyberman opened the pull request:
I would like to suggest merging of the following two commits:

**Commit aac74bce7e22**

https://github.com/biocyberman/piper/commit/aac74bce7e22137e6fb7a539f21102c7ef5ebf3e

I created a 'config' directory at project base and moved three original files to 'config/upsala': `globalConfig.sh globalConfig.xml uppmax_global_config.xml`. I then modified `setup.sh` file to allow this syntax: 

`./setup.sh /path/to/install/piper institution /path/to/otherlibraries/lib NoGATKCompile`

An example command is like this: `./setup.sh /space/system/piper afmd /space/system/lib nogatk`

Which means:

1. To install piper to ` /space/system/piper`
2. To pickup config files under `config/afmd` directory. Current possible choices are `afmd` and `upsala`, but more subdirectories for any new institutions can be added.
3. Register /space/system/lib to piper so it can look for additional library dependencies (If they are put there, of course)
4. The fourth argument, whenever none-empty, will tell `setup.sh` not to download and compile GATK. This is nicer than commenting/uncommenting the function call.


**Commit 704a1ac9d26cc686d** 
This is an additional commit to remove dependency on `GenomeAnalysisTK.jar`. This is because `Queue.jar` already includes content of `GenomeAnalysisTK.jar`. Putting both of them together in `lib` causes warning about redundant packages found, such as the one about `org.slf4j` bindings.

https://github.com/biocyberman/piper/commit/704a1ac9d26cc686d460fd93dc9b551f6ced77f5

My commit tree looks a bit messy, I will try to keep them in good order from now on :)

johandahlberg commented:
Hi!

Awesome PR. Unfortunately I don't have time to review it just now. But I've done a quick read through, and I think that it looks good. Until I have time to look at this more thoroughly maybe you could get the travis build working? As you will see (here)[https://travis-ci.org/NationalGenomicsInfrastructure/piper/jobs/50278087] it's complaining about not finding some dependencies (scroll to the end of the build log).

Thanks for the PR! :smile:

biocyberman commented:
Hi Johan
travis failed because it JHDF5 library to deal with XSQ colorspace data was not uploaded. I will fix that. However, I am thinking about logistics problem. Would it be better to submit PRs to master branch so we can 'communicate' through this common ground? Stable releases can be done via release tags that you have been doing or  on a 'stable' branch. **If you agree I will close this PR and open a new one to the master branch.**  PLEASE GIVE YOUR COMMENT ON THIS.

And some more words about the QPipe branch, even though I try to keep the 'upstream' intact, I had to make some changes to simplify the implementation of QPipe branch: To support XSQ and maybe BAM inputs later. I've made some rather fundamental changes in addtion to the one described in the original PR:

1. Rewrite ReadPairContainer class and refactor it to `InputSeqFileContainer`. Follow this I refactored all uses of `ReadPairContainer` in the repo. Most of the cases it is just a matter of renaming because  `InputSeqFileContainer` is backward-compatible. `InputSeqFileContainer` supports more flexible and more generic ways to deal with different input formats.
2. Configure logging to work properly. However, this also require to upgrade org.slf4j in Queue to at least 1.7.5 version. So, `setup.sh` will NOT work if it enables compiling of GATK. An exception of `MethoNotFound` will occur because of the outdated version of SLF4J embedded in Queue/GATK. The best way is that I will ask to upgrade SLF4J in Queue/GATK at its origin.   
3. Upgrade scala to 2.10.4 and sbt to 0.13.7. 
4. Upgrade `commons-lang`, `commons-io` package.   This is because somehow these two packages in Piper's SBT build intefer with the same package names in Queue/GATK, causing again `MethodNotFound` exception when Queue tries to use `FileUtils.FileCopy` method. 

I hope will these changes to be merged into the master branch of Piper. Since it was sudden jump for me to Piper, I had to do several experiments and therefore forgot to commit all changes on time. This results in not so well isolated commits. I will anyways, do a 'clean-up' commit and start with well isolated ones. Back in the head, I think the code base is not too complicated, so it is not to bad to do this clean-up commit. This may challenge your patience :-)

biocyberman commented:
Hi Johan,
I have just pushed some new commits to QPipe branch. Notably in commit 894e8ac, I converted the project to Maven-based, which allows straight forward integration with GATK/Queue. This is important because it will be easy to refer to Queue source or including dependency directly from github Maven-based repos (i.e. gatk-projected repo).

johandahlberg commented:
Hi!

I'm just dropping a note here to let you know that I've not forgotten this. However I've been bogged down with a lot of other stuff to do lately and I've haven't had the time to look into this. I'll promise to get to it as soon as time permits.

/Johan

biocyberman commented:
No worries Johan, I am in a not-so-different situation.

johandahlberg commented on .gitignore:
```diff
@@ -1,5 +1,10 @@
 target/
+Queue.jar
+GenomeAnalysisTK.jar
```
Shouldn't this be covered by the *.jar clause?

biocyberman replied:
There is an dependency in lib/sis-jhdf5-batteries_included.jar that can't be downloaded from Maven central repository. It will be ignored if we use *.jar. Anyway, I have a fix for it by downloading and preparing it from jhdf5 homepage.

johandahlberg commented on config/afmd/globalConfig.sh:
```diff
@@ -0,0 +1,140 @@
+# Start by exporting the shared drmaa libaries to the LD_LIBRARY_PATH
```
Just dropping a comment here to let you know that my intention is that in the future this file should be removed and all config should instead be moved into the xml files, to not spread it all over the place. But since some workflows still rely on this I haven't gotten rid of it yet.

biocyberman replied:
I got this impression somewhere while reading the code. Some GLOBAL variables are still defined on this script. The fix is simple. We will come to this later.

johandahlberg commented on src/main/scala/molmed/apps/MakeSolidProjectHierarchy.scala:
```diff
@@ -0,0 +1,322 @@
+  /**
+   * List all subdirectories of dir
+   */
+  def listSubDirectories(dir: File): Seq[File] = {
```
I guess that this type of function could be refactored out and placed in some general companion object.

johandahlberg replied:
Looking at all of this code I'm guessing that that goes for a lot of this stuff most of it seems to be copy-pasted from the original class.

biocyberman replied:
Yes, I first started with duplicating the file/class closest to the one I want to write and start making changes therein. It is a lazy and error-prone process, but I since any file with name containing "Solid" is ONLY to deal with colorspace data. So the risk is only on my shoulder, they do not affect the upstream code. With that said, this adaptation work is still going on and I do improve them quickly.

johandahlberg commented on src/main/scala/molmed/apps/SolidGenerateXMLfromCSV.scala:
```diff
@@ -0,0 +1,80 @@
+    val seqFiles = line.drop(4)
+
+
+    // Sometime the meta data file contains more files than the folder...
```
Maybe remove this block instead of commenting it out?

biocyberman replied:
Good idea. However, currently I do not use this class so I will improve when there is an obvious need to make it work.

johandahlberg commented on src/main/scala/molmed/apps/setupcreator/SolidSetupUtils.scala:
```diff
@@ -0,0 +1,433 @@
+
+
+//
+//trait SetupUtils{
```
Remove comment block?

johandahlberg commented on src/main/scala/molmed/config/FileAndProgramResourceConfig.scala:
```diff
@@ -47,12 +47,15 @@ trait FileAndProgramResourceConfig {
-  @Input(doc = "The path to the binary of bwa (usually BAM files have already been mapped - but if you want to remap this is the option)", fullName = "path_to_bwa", shortName = "bwa", required = false)
-  var bwaPath: File = _
+  @Input(doc = "The path to the binary of bwa (usually BAM files have already been mapped - but if you want to remap this is the option)", fullName = "path_to_aligner", shortName = "alner", required = false)
+  var alignerPath: File = _
```
Might be better to keep this as `bwaPath` since as things are generally setup here they set it as the path to a specific program?

johandahlberg commented on src/main/scala/molmed/config/FileAndProgramResourceConfig.scala:
```diff
@@ -102,87 +117,15 @@ trait FileAndProgramResourceConfig {
     doNotLoadDefaultResourceFiles: Boolean = false): ResourceMap = {
 
     /**
-     * This trait is used to make sure that both programs
```
Yeah. It makes sense to move this.

johandahlberg commented on src/main/scala/molmed/config/FileAndProgramResourceConfig.scala:
```diff
@@ -263,6 +209,15 @@ trait FileAndProgramResourceConfig {
       if (this.snpEffPath == null)
         this.snpEffPath = getFileFromKey(programNameToPathsMap, Constants.SNP_EFF)
 
+      if (this.novoalignCS == null)
```
I'm not sure, but maybe this should be moved into a separate trait. If you are planing to use write your own QScripts that would create more separation and less confusion for the end user since they don't have to be presented with options for lots of programs that they are not going to use..

biocyberman replied:
Good suggestion. And the comment about bwaPath is very reasonable. I did that changed before I decided to introduce novoalignCS in the xml file. But, again you are right, it is a good idea to move them Colospace things to a trait because right now they cause troubles to 'upsala' config.

johandahlberg commented on src/main/scala/molmed/queue/setup/InputSeqFile.scala:
```diff
@@ -0,0 +1,44 @@
+ * @param fileType  What data format are the input sequence files. Possible values are FASTQ, BAM, or XSQ. Default to FASTQ
+ */
+case class InputSeqFileContainer( files: Seq[File],
+                                  var sampleName: String = null,
```
Would be cleaner I think to replace `null` with using `Option`. I know that I'm the one responsible for this rather ugly construct from the beginning though.

biocyberman replied:
I agree with using `Option`. How do I mark this for later refactoring? It is not an urgent matter for this so I don't think we need to do it before others.

johandahlberg replied:
Well if it's something this small I normally just drop a "TODO" comment in the code - but you could also report it as an issue.

johandahlberg commented on src/main/scala/molmed/report/ReportGenerator.scala:
```diff
@@ -63,34 +63,36 @@ object ReportGenerator {
+//      version
+//    } else
+//      Constants.unknown
+    org.broadinstitute.gatk.engine.CommandLineGATK.getVersionNumber
```
Well. That was a lot cleaner. :grinning:. Maybe lose all the commented code above and keep you **much** simpler solution.

biocyberman replied:
Hehe, I found it by chance when I was trying to remove dependency for GenomeAnalysisTK.jar. The key is that I knew we don't kneed Queue.jar and GenomeAnalaysisTK..jar to run `piper`.

johandahlberg commented on src/main/scala/molmed/utils/GATKUtils.scala:
```diff
@@ -79,6 +79,7 @@ class GATKUtils(gatkOptions: GATKConfig, projectName: Option[String], uppmaxConf
     this.covariate ++= Seq("ReadGroupCovariate", "QualityScoreCovariate", "CycleCovariate", "ContextCovariate")
     this.input_file :+= inBam
     this.disable_indel_quals = false
+    this.solid_nocall_strategy = org.broadinstitute.gatk.utils.recalibration.RecalUtils.SOLID_NOCALL_STRATEGY.PURGE_READ
```
I think this needs to be made optional and only activated for the solid case.

biocyberman replied:
You are right. I made changes to reflex that.

johandahlberg commented on src/main/scala/molmed/utils/VariantCallingUtils.scala:
```diff
@@ -433,13 +433,13 @@ class VariantCallingUtils(gatkOptions: GATKConfig, projectName: Option[String],
     //  From best practice: -an DP -an QD -an FS -an MQRankSum -an ReadPosRankSum
     this.use_annotation ++= List("QD", "FS", "MQRankSum", "ReadPosRankSum")
 
+    this.resource :+= new TaggedFile(gatkOptions.hapmap.get, "known=false,training=true,truth=true,prior=15.0")
```
Good call! :+1:

johandahlberg commented on src/test/scala/molmed/apps/MyLoggers.scala:
```diff
@@ -0,0 +1,27 @@
+package molmed.apps
```
I suppose this is some kind of test code that could be removed?

biocyberman replied:
Yes, it will be. I is a dirty quick test when I was having a headache with log4j and slf4j conflict.

johandahlberg commented on src/test/scala/molmed/apps/TestLinkXSQFiles.scala:
```diff
@@ -0,0 +1,24 @@
+package molmed.apps
```
Is this also something that could be removed?

biocyberman replied:
This will be rewritten to be more like a test.

johandahlberg commented:
I've now had a quick read through of the code. and you can see my comments on it inline.

Some things of more general node:
* I agree that it's a good idea to create a common stable release branch to which future pull requests can then be created.
* When it comes to the setup process I have been working on moving it to a rpm-based process which should make things considerably easier to deploy. Then you could use environment variables to specify your particular workflow dir (and possibly even move those things into their own repository). It feels a bit clunky that one would have to add a a separate folder structure for every place that this should be installed.
* When it comes to switching to a maven based life cycle I feel somewhat reluctant since my experience with maven is limited, and what the little experience I've have with it I've found the xml-style to be uncomfortable to work with. However, I'm open to being persuaded if you have strong arguments for switching. When it comes to being able to integrated directly with maven-based github-repos such as the GATK I found another solution some time ago which I found very interesting but that I haven't had the time to try out yet. The idea would be to use [jitpack.io](https://jitpack.io/) to pull down the Queue dependencies.

biocyberman commented:
Hi Johan,
Thanks for taking time to look at the PR. Regarding the last two points:

* "When it comes to the setup process..": I haven't thought about rpm-based. Would it be a step to limit deb-base user like Ubuntu, Debian? I also felt the difficulty of current build for deployment. I went a head and create an standalone installable .sh file with binary payload. I will commit and let you see the code. With the current suggestion of having specific folder for each center under `config`, I thought that not every end-user should commit back their center specific files. In additions, we can easily parameterize it to pickup a folder outside git repo at build time. That is for the wrapper bash scripts. For xml files, it may be a good idea to create a bash script or an app to read user input and write the files.
*  " switching to a maven based life cycle..." Maven was very intimidating to me at first when I looked at the POM files of Gatk/Queue project. After some effort of reading two very sonatype books: http://books.sonatype.com/mvnex-book/reference/public-book.html and watching some youtube clips, I started experimenting and see that it is a very logical and predictable process to work with POM files (with the help of Intellij/Eclipse). Better integration with Queue is a very nice thing, but not the most important factor. The main drive that brought me to maven is the ability to control very well dependencies with "high resolution" AND do all-sort of manipulation with the content of final binary jar, javadoc, and sources jar files. As you can see I already merged every jar file that `piper` needs and piper itself in or jar file. It maybe to few details to persuade you for Maven-based. Anyways, I think all usage scenarios of Maven for `Piper` have been addressed in the POM file. So, reluctant or not, you have next to none of headache left to worry about Maven. Of course, there are many great other uses with Maven in other project if you feel comfortable with it.

I will come back for some more commits later to get you updated with the latest.

biocyberman commented:
Hi Johan
I addressed most of the points you commented in the previous commits. As you can see only some non-essential points are left. When all are done, just tell me if you want me to create a PR to the master branch. Otherwise you can just merge to the master branch by yourself.

johandahlberg commented:
Hi @biocyberman! I've been feeling guilt about not addressing this for a very long time. I'm very sorry for leaving this hanging here.

The scope of Piper has drifted a bit over time, and now it's mostly being used to power our human whole genome sequencing pipelines - which means that at this point I'm skeptical about it being worth the effort for anyone else to go into the project. I'm going to close this pull request now - but if you feel like you are still interested in extending Piper or if you would like me to put you into contact with some other folks I know who seem to me to be working on things which are more easily adaptable to different centers, send me an e-mail at johan.dahlberg@medsci.uu.se and we can have a chat about that.

Once again I'm so sorry for not coming back to you on this, much, much, sooner.
